python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -503,10 +508,16 @@ def get() -> "DagsterInstance":
     @public
     @staticmethod
     def local_temp(
-        tempdir: Optional[str] = None, overrides: Optional[DagsterInstanceOverrides] = None
+        tempdir: Optional[str] = None,
+        overrides: Optional[DagsterInstanceOverrides] = None,
     ) -> "DagsterInstance":
         if tempdir is None:
-            tempdir = DagsterInstance.temp_storage()
+            created_dir = TemporaryDirectory()
+            i = DagsterInstance.from_ref(
+                InstanceRef.from_dir(created_dir.name, overrides=overrides)
+            )
+            DagsterInstance._TEMP_DIRS[i] = created_dir
```

Comment:
I briefly went  down this path for the reasoning you laid out but due to way the `local_temp` instance is currently constructed the refactor to accomplish that would be larger than i thought worth it.

Reply:
A version that wouldn't be a big of a refactor is making it so that `DagsterInstance` has an `ExitStack` and exposes tacking things on to it, `instance.add_context` or something. Would you prefer that?